lib/Dialect/RTL/RTLFolds.cpp

```diff
@@ -706,3 +715,289 @@ void ConcatOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::sge:
+    if (value.isMinSignedValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::ult:
+    if (value.isMinValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::ule:
+    if (value.isMaxValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::ugt:
+    if (value.isMaxValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::uge:
+    if (value.isMinValue())
+      return getTrueAttr(context);
+    return {};
+  default:
+    return {};
+  }
+}
+
+OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
+
+  // gt a, a -> false
+  // gte a, a -> true
+  if (lhs() == rhs()) {
+    auto val = applyCmpPredicateToEqualOperands(predicate());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  auto lhs = constants.front().dyn_cast_or_null<IntegerAttr>();
+  auto rhs = constants.back().dyn_cast_or_null<IntegerAttr>();
+
+  // gt 1, 2 -> false
+  if (lhs && rhs) {
+    auto val = applyCmpPredicate(predicate(), lhs.getValue(), rhs.getValue());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  // lt x, 0 -> false
+  if (rhs)
+    return tryApplyCmpPredicateToRHS(predicate(), rhs);
+  else if (lhs)
+    return tryApplyCmpPredicateToRHS(flipPredicate(predicate()), lhs);
+
+  return {};
+}
+
+void ICmpOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+                                         MLIRContext *context) {
+  struct Folder final : public OpRewritePattern<ICmpOp> {
+    using OpRewritePattern::OpRewritePattern;
+    LogicalResult matchAndRewrite(ICmpOp op,
+                                  PatternRewriter &rewriter) const override {
+
+      APInt lhs, rhs;
+
+      // icmp 1, x -> icmp x, 1
+      if (matchPattern(op.lhs(), m_RConstant(lhs)) &&
+          !matchPattern(op.rhs(), m_RConstant(rhs))) {
+        rewriter.replaceOpWithNewOp<ICmpOp>(op, flipPredicate(op.predicate()),
+                                            op.rhs(), op.lhs());
+        return success();
+      }
+
+      // gt x, max-1 -> eq x, max
+      // lt x, min+1 -> eq x, min
+      if (matchPattern(op.rhs(), m_RConstant(rhs))) {
+        ConstantOp constant;
+        switch (op.predicate()) {
+        case ICmpPredicate::slt:
+          if ((rhs - 1).isMinSignedValue())
+            constant = rewriter.create<ConstantOp>(op.getLoc(), rhs - 1);
+          break;
+        case ICmpPredicate::sgt:
+          if ((rhs + 1).isMaxSignedValue())
+            constant = rewriter.create<ConstantOp>(op.getLoc(), rhs + 1);
+          break;
+        case ICmpPredicate::ult:
+          if ((rhs - 1).isMinValue())
+            constant = rewriter.create<ConstantOp>(op.getLoc(), rhs - 1);
+          break;
+        case ICmpPredicate::ugt:
+          if ((rhs + 1).isMaxValue())
+            constant = rewriter.create<ConstantOp>(op.getLoc(), rhs + 1);
```

Comment:
I'd recommend handling this by canonicalizing `x<=cst` -> `x < cst+1`

Reply:
This is handling the weird middle case in this chain: `x <= min` -> `x < min+1` -> `x == min`.  Should I handle this whole chain differently?

Reply:
I'd recommend you merge all the "match rhs against constant" cases into a single switch.  Something like this:

switch (op.predicate()) {
...
   case ICmpPredicate::ugt:
          if (rhs.isMaxValue()) {
            resultConstant = getfalseconstant; 
            break;
          }
          // Strength reduce to ==
          if ((rhs + 1).isMaxValue()) {
            equalConstant = rewriter.create<ConstantOp>(op.getLoc(), rhs + 1);
            break;
          }
        same for notEqualConstant
        

In the `<=` cases, handle them by just converting to `<` and let all the machinery above do the right thing.  This means you just have to check the maxint case (which sets resultConstant) and otherwise just turns it into a "x < c+1" comparison.

at the bottom of the switch you can have:

if (resultConstant) { replaceOpWith(op, resultConstant); return success();}
if (equalConstant) { replace with icmp equal of that constant }
etc.

The key with this kind of peephole optimization is to make it as simple and short as possible, because there are tons of special cases that will be accreted over time and they are all low probability of happening in any given call.

Reply:
I used a couple helper functions to handle it inside the switch, instead of resolving the action at the end (`if(equalConstant) { ... {`)